File: tests/wasm2lang_13_i64_casts.harness.mjs
'use strict';

const moduleImports = {
  i64_to_f32: x => Math.fround(Number(x)),
  i64_to_f64: x => Number(x),
  f32_to_i64: x => BigInt(Math.trunc(x)),
  f64_to_i64: x => BigInt(Math.trunc(x))
};

const wasmImports = {cast: moduleImports};

const runTest = function (buff, out, exports, data) {
  exports.alignHeapTop();

  for (const t of data.cast_triples) {
    exports.exerciseI64Casts(t[0], Math.fround(t[1]), t[2]);
  }

  exports.exerciseI64CastEdgeCases();
};

const dumpMemory = true;

export {dumpMemory, moduleImports, runTest, wasmImports};

File: tests/wasm2lang_13_i64_casts.harness.java
// Java test harness for wasm2lang_13_i64_casts.
{
    java.util.Map<String, Object> foreign = new java.util.LinkedHashMap<>();
    foreign.put("i64_to_f32", (java.util.function.LongToDoubleFunction)(long x) -> (float) x);
    foreign.put("i64_to_f64", (java.util.function.LongToDoubleFunction)(long x) -> (double) x);
    foreign.put("f32_to_i64", (java.util.function.DoubleToLongFunction)(double x) -> (long) x);
    foreign.put("f64_to_i64", (java.util.function.DoubleToLongFunction)(double x) -> (long) x);

    WasmModule mod = new WasmModule(foreign, memBuffer);
    mod.alignHeapTop();

    java.util.Map<String, Object> _data = w2lLoadSharedData(System.getProperty("w2l.testname", ""));

    for (java.util.List<Double> t : w2lNested(_data, "cast_triples")) {
        mod.exerciseI64Casts(t.get(0).intValue(), t.get(1).floatValue(), t.get(2));
    }

    mod.exerciseI64CastEdgeCases();

    w2lDumpCRC(memBuffer);
}

/exit
